contracts/upkeeps/Cron.sol

```diff
@@ -0,0 +1,294 @@
+ * Users must use the encodeCronString() function to encode their cron jobs before
+ * setting them. This keeps all the string manipulation off chain and reduces gas costs.
+ */
+contract Cron is KeeperCompatibleInterface, ConfirmedOwner, Pausable {
+  using CronUtility for Spec;
+  using CronUtility for string;
+
+  event CronJobExecuted(uint256 indexed id, uint256 timestamp);
+  event CronJobCreated(uint256 indexed id, address target, bytes handler);
+  event CronJobDeleted(uint256 indexed id);
+
+  error CallFailed(uint256 id, string reason);
+  error CronJobIDNotFound(uint256 id);
+  error InvalidHandler();
+  error TickInFuture();
+  error TickTooOld();
+  error TickDoesntMatchSpec();
+
+  uint256 private s_nextCronJobID = 1;
+  uint256[] private s_activeCronJobIDs;
+
+  mapping(uint256 => uint256) private s_lastRuns;
+  mapping(uint256 => Spec) private s_specs;
+  mapping(uint256 => address) private s_targets;
+  mapping(uint256 => bytes) private s_handlers;
+  mapping(uint256 => bytes32) private s_handlerSignatures;
+
+  constructor() ConfirmedOwner(msg.sender) {}
+
+  /**
+   * @notice Executes the cron job with id encoded in performData
+   * @param performData abi encoding of cron job ID and the cron job's next run-at datetime
+   */
+  function performUpkeep(bytes calldata performData)
+    external
+    override
+    whenNotPaused
+  {
+    (uint256 id, uint256 tickTime, address target, bytes memory handler) = abi
+      .decode(performData, (uint256, uint256, address, bytes));
+    validate(id, tickTime, target, handler);
+    s_lastRuns[id] = block.timestamp;
+    (bool success, bytes memory payload) = target.call(handler);
+    if (!success) {
+      revert CallFailed(id, getRevertMsg(payload));
+    }
+    emit CronJobExecuted(id, block.timestamp);
+  }
+
+  /**
+   * @notice Creates a cron job from the given encoded spec
+   * @param target the destination contract of a cron job
+   * @param handler the function signature on the target contract to call
+   * @param encodedCronSpec abi encoding of a cron spec
+   */
+  function createCronJobFromEncodedSpec(
+    address target,
+    bytes memory handler,
+    bytes memory encodedCronSpec
+  ) external {
+    Spec memory spec = abi.decode(encodedCronSpec, (Spec));
+    createCronJobFromSpec(target, handler, spec);
+  }
+
+  /**
+   * @notice Deletes the cron job matching the provided id. Reverts if
+   * the id is not found.
+   * @param id the id of the cron job to delete
+   */
+  function deleteCronJob(uint256 id) external onlyOwner {
+    if (s_targets[id] == address(0)) {
+      revert CronJobIDNotFound(id);
+    }
+    uint256 existingID;
+    uint256 oldLength = s_activeCronJobIDs.length;
+    uint256 newLength = oldLength - 1;
+    for (uint256 idx = 0; idx < oldLength; idx++) {
+      existingID = s_activeCronJobIDs[idx];
+      if (existingID == id) {
+        s_activeCronJobIDs[idx] = s_activeCronJobIDs[newLength];
+        break;
+      }
+    }
```

Comment:
```suggestion
uint256 idx;
    for (idx = 0; idx < newLength; idx++) {
      existingID = s_activeCronJobIDs[idx];
      if (existingID == id) {
        s_activeCronJobIDs[idx] = s_activeCronJobIDs[newLength];
        break;
      }
    }
require(idx < oldLength, "id not found");
```
I didn't realize that this version does not handle non-inclusion check that yours had. Open to going back to the old way instead of this change.